azkaban-db/src/main/java/azkaban/db/DatabaseOperator.java

```diff
@@ -117,16 +118,35 @@ public DatabaseOperator(final QueryRunner queryRunner) {
    * @return The number of rows updated.
    */
   public int update(final String updateClause, final Object... params) throws SQLException {
-    try {
-      return this.queryRunner.update(updateClause, params);
-    } catch (final SQLException ex) {
-      // todo kunkun-tang: Retry logics should be implemented here.
-      logger.error("update failed", ex);
-      if (this.dbMetrics != null) {
-        this.dbMetrics.markDBFailUpdate();
+    int retry = 0;
+    SQLException exception;
+    String errorMsg = "Update failed: Reached maximum number of retries: " + AzDBUtil.MAX_RETRIES_ON_DEADLOCK;
+    do {
+      try {
+        return this.queryRunner.update(updateClause, params);
+      } catch (final SQLException ex) {
+        exception = ex;
+        if (ex.getErrorCode() == 1213) { // 1213 -> Deadlock error
+          retry++;
+          logger.warn("Deadlock detected when trying to execute: " + updateClause + " with values: "
+              + Arrays.toString(params));
+          try {
+            Thread.sleep(AzDBUtil.RETRY_WAIT_TIME);
+          } catch (InterruptedException e) {
+            logger.info("Sleep interrupted");
```

Comment:
A more detailed message like "Sleep during DB update retry interrupted" would be more useful.

Reply:
Would it be better if the actual SQL is also logged? This way, we can learn about which queries get in this kind of situation and how often.

Reply:
In my opinion there is no relation between the sleep being interrupted and the query being executed, it's just a waiting time that was not honored. Adding the query to the logs would only cause confusion and not add interesting info IMO.
The query should certainly be logged if it fails.